nntrainer/src/conv2d_layer.cpp

```diff
@@ -138,6 +182,91 @@ sharedConstTensor Conv2DLayer::backwarding(sharedConstTensor derivative,
     delBias.setZero();
   }
 
+  Tensor ret(input_dim.batch(), input_dim.channel(),
+             input_dim.height() + padding[0] * 2,
+             input_dim.width() + padding[1] * 2);
+  ret.setZero();
+
+#ifdef USE_BLAS
+  // Calculate DelK, DelBias
+  int status = ML_ERROR_NONE;
+  for (unsigned int b = 0; b < input_dim.batch(); ++b) {
+    std::vector<float> out(kernel_size[0] * kernel_size[1] *
+                           input_dim.channel() * filter_size);
+
+    Tensor inSub(
+      TensorDim(1, input_dim.channel(), input_dim.height(), input_dim.width()),
+      input.getAddress(b * input.getDim().getFeatureLen()));
+
+    status = conv2d_gemm(
+      derivative->getAddress(b * derivative->getDim().getFeatureLen()),
+      TensorDim(1, derivative->channel(), derivative->height(),
+                derivative->width()),
+      inSub,
+      TensorDim(1, 1, filter_size,
+                kernel_size[0] * kernel_size[1] * input_dim.channel()),
+      stride, padding, out.data(), out.size(), false);
+    if (status != ML_ERROR_NONE)
+      throw std::runtime_error("Backwarding Convolution failed.");
+
+    for (unsigned int i = 0; i < filter_size; ++i) {
+      Tensor &delK = paramsAt(i).grad;
+      Tensor &delBias = paramsAt(i + filter_size).grad;
+      float *del = delK.getData();
+      unsigned int s = kernel_size[0] * kernel_size[1] * input_dim.channel();
+
+      for (unsigned int k = 0; k < s; ++k) {
+        del[k] += out[i * s + k];
+      }
+
+      float sum = 0.0;
+      for (unsigned int j = 0; j < derivative->height(); ++j) {
+        for (unsigned int k = 0; k < derivative->width(); ++k) {
+          sum += derivative->getValue(b, i, j, k);
+        }
+      }
+      delBias.setValue(0, 0, 0, 0, sum + delBias.getValue(0, 0, 0, 0));
+    }
+  }
+
+  // Calculate Derivative to propagation
+  TensorDim kdim(ret.channel(), filter_size, kernel_size[0], kernel_size[1]);
+
+  std::vector<float> imKernel(kdim.getDataLen());
+
+  unsigned int count = 0;
+  float *d = imKernel.data();
+
+  for (unsigned int j = 0; j < ret.channel(); ++j) {
+    for (unsigned int i = 0; i < filter_size; ++i) {
+      Tensor &filters = paramsAt(i).weight;
+      for (unsigned int k = 0; k < kernel_size[0] * kernel_size[1]; ++k) {
+        d[count++] = filters.getData()[j * kernel_size[0] * kernel_size[1] + k];
+      }
+    }
+  }
+
+  TensorDim input_dim_padded(1, input_dim.channel(),
+                             input_dim.height() + padding[0] * 2,
+                             input_dim.width() + padding[1] * 2);
+
+  for (unsigned int b = 0; b < input_dim.batch(); ++b) {
+    Tensor inSub(
```

Comment:
This constructor does memcpy. I think no need to memcpy here if conv2d_gemm takes argument of const float * in instead of const const Tensor & in

Reply:
the input of conv2d_gemm is tensor which is with dimension and it is used to calculate padding.